### PR TITLE
emoji/unsafe identifiers and identifier collisions

### PR DESCRIPTION
**Pull request by skibz**
[as discussed](https://github.com/lieuwex/whapp-irc/issues/45) previously:

* [x] emojis found in chat identifiers are converted to emojitag, and then have their unsafe characters stripped using the existing regex solution
* [x] identifiers that have already been encountered are given an integer suffix to make them distinct
* [x] adds test (`util_test.go`) to assert behaviour is correct
* [x] adds a dependency (😭) to help with managing emoji 

* [x] add missing empty string test cases
* [x] add test cases for other types of invalid irc nicknames (eg: `$$$`)
* [x] implement an encoding solution for said nicknames

### CODE REVIEW

**Comment by skibz** [comment]
i just realised now while looking at the diff that i should not have exported `ircSafeString` because there is already `SafeString` exported. sorry about that! i will fix this up as soon as possible.

**Comment by skibz** [comment]
it *also* just occurred to me that the tests i've written never assert that a resulting identifier is not an empty string! which is totally missing the point of this whole exercise 😄 

i should also have a test case for when an identifier is entirely composed of non-emoji special characters.

what i'm thinking is that if someone has the name `$$$` in whatsapp, they'd also end up with an empty string for name. so ultimately we'd have to find a new encoding to use for those sorts of names.

perhaps we could convert those names to a string containing the original identifier's hexadecimal representation, then prefix it with `x` so that it doesn't violate irc nickname syntax (a name cannot start with number, if i recall)

eg: using the original example name `$$$` the encoding i'm proposing would result in the name `x242424`.

a client can then also decode the name back to `$$$` quite easily. to be clear, i propose hex because i can't think of an encoding that doesn't violate irc name rules. if you know of another, please indicate.

File: util.go [hunk 1]
```diff
@@ -51,11 +58,31 @@ func getExtensionByMimeOrBytes(mime string, bytes []byte) string {
...
+	return distinct
+}
+
+func ensureIdentifierIsDistinct(identity string) string {
+	_, exists := identifiers[identity]
+
+	// we've encountered this identifier before so
+	// increment the counter and append the new count
+	// to the identifier we return
+	if exists {
```
Comment (lieuwex) [inline util.go]:
```suggestion
	if _, exists := identifiers[identity]; exists {
```

and removing the line from 71-72, of course

File: util.go [hunk 2]
```diff
@@ -51,11 +58,31 @@ func getExtensionByMimeOrBytes(mime string, bytes []byte) string {
 	return getExtension(bytes)
 }
 
-var unsafeRegex = regexp.MustCompile(`(?i)[^a-z\d+]`)
+// IrcSafeString converts any emoji unicode characters into emojitag, then converts any non-ascii characters into their ascii equivalents, then strips characters that satisfy unsafeRegex, and finally disambiguates the identifier if required
+func IrcSafeString(str string) string {
+	emojiTagged := emoji.UnicodeToEmojiTag(str)
+	decoded := unidecode.Unidecode(emojiTagged)
+	ircSafe := unsafeRegex.ReplaceAllLiteralString(decoded, "")
+	distinct := ensureIdentifierIsDistinct(ircSafe)
```
Comment (lieuwex) [inline util.go]:
```suggestion
	return ensureIdentifierIsDistinct(ircSafe)
```

File: util.go [hunk 3]
```diff
@@ -51,11 +58,31 @@ func getExtensionByMimeOrBytes(mime string, bytes []byte) string {
 	return getExtension(bytes)
 }
 
-var unsafeRegex = regexp.MustCompile(`(?i)[^a-z\d+]`)
+// IrcSafeString converts any emoji unicode characters into emojitag, then converts any non-ascii characters into their ascii equivalents, then strips characters that satisfy unsafeRegex, and finally disambiguates the identifier if required
```
Comment (lieuwex) [inline util.go]: Please limit line length to 80 for comments like these.

File: util_test.go [hunk 4]
```diff
@@ -0,0 +1,77 @@
...
+	simpleEmojiStr := emoji.EmojiTagToUnicode(":ok_hand:" + unsafeCharacters)
+
+	// we invoke the test subject a few times in a
+	// row to test the map[string]int side effects
+	simpleSafeStr1 := IrcSafeString(simpleEmojiStr)
+	simpleSafeStr2 := IrcSafeString(simpleEmojiStr)
+	simpleSafeStr3 := IrcSafeString(simpleEmojiStr)
+
+	// ensure none of the strings match
+	if simpleSafeStr1 == simpleSafeStr2 || simpleSafeStr1 == simpleSafeStr3 || simpleSafeStr2 == simpleSafeStr3 {
```
Comment (lieuwex) [inline util_test.go]: Also these test lines can be split nicely:

```suggestion
	if simpleSafeStr1 == simpleSafeStr2 ||
	simpleSafeStr1 == simpleSafeStr3 ||
	simpleSafeStr2 == simpleSafeStr3 {
```

File: util_test.go [hunk 5]
```diff
@@ -0,0 +1,77 @@
...
+	// ensure the duplicate strings have been made distinct with an incrementing integer
+	if !strings.HasSuffix(simpleSafeStr3, "3") {
+		t.Fatalf("expected duplicate simple emoji string to end with a disambiguating numeric suffix: %s", simpleSafeStr3)
+	}
+
+	t.Logf("simple emoji strings are unique after invoking IrcSafeString: %s %s %s", simpleSafeStr1, simpleSafeStr2, simpleSafeStr3)
+}
+
+func TestIrcSafeStringComplex(t *testing.T) {
+
```
Comment (lieuwex) [inline util_test.go]: whitespace at the beginning and end of this function

**Comment by skibz** [comment]
> The problem with appending a number at the end of the identifier is is that the order of the chats in memory is arbitrary, so after a restart the identifier<->chat mapping can be completely different, or not, it depends. We should maybe add some ordering to the chat list as well.

oh man, i hadn't considered this. what sort of ordering do you propose? maybe there's a way we can side-step this problem entirely. is there a data-point provided by whatsapp that could be used to uniquely identify contacts?

**Comment by lieuwex** [comment]
Okayyyy, due to some difficulties in how I wanted to fix the problem, I have merged your PR as 6732fee136bb05313abcad648ddf974aab790b13. The tests are missing, I don't really care, but if you do you can rewrite them since the code is a bit changed now :P. Thanks!

**Comment by skibz** [comment]
🍾
